base/inc/AdePT/SparseVector.h

```diff
@@ -0,0 +1,746 @@
+    //  - ballot the requests coming from threads of the same warp
+    //  - count the requests and let one thread do the atomic increment and get the base offset before incrementing
+    //  - broadcast the offet to all threads in the warp
+    //  - compute the index to be assigned to each request
+    // TODO: use the `Opportunistic Warp-level Programming` approach from:
+    //  https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/
+    set_bit(index);
+    return (data() + index);
+  }
+
+  /** @brief Release the element at index i */
+  __host__ __device__ __forceinline__ void release(Index_t i)
+  {
+    // Make sure the element is not already released
+    assert(is_used(i) && "Trying to release an element not in use.");
+    fNused--;
+    reset_bit(i);
+  }
+
+  /** @brief Check if the element with index i is used */
+  __host__ __device__ __forceinline__ bool is_used(Index_t i) const
+  {
+    assert(i < fCapacity);
+    return (mask_at(i / 32).load() & (1 << (i & 31)));
+  }
+
+  /** @brief Check if container is fully distributed */
+  __host__ __device__ __forceinline__ bool is_full() const { return (size() == fCapacity); }
+
+  /** @brief Check if container is empty */
+  __host__ __device__ __forceinline__ bool is_empty() const { return (size_used() == 0); }
+
+  /** @brief Fills selection vector fSelected with element indices passing the user predicate
+   *  @param pred Predicate function taking as arguments the SparseVector pointer and the element index
+   *  @returns Number of selected elements. Must be reset to 0 when calling the function.
+   */
+#ifdef COPCORE_CUDA_COMPILER
+  template <typename Predicate, typename Container>
+  static void select(const Vector_t *svect, Predicate pred_func, Container *output, unsigned *num_selected)
+  {
+    sa_detail::select_kernel_launcher<Type, Predicate, Container>
+        <<<1, 1>>>((Vector_t *)svect, pred_func, output, num_selected);
+  }
+#else
+  template <typename Predicate, typename Container>
+  static void select(const Vector_t *svect, Predicate pred_func, Container *output, unsigned *num_selected)
+  {
+    // host version
+    const int num_items = svect->size();
+    *num_selected       = 0;
+    for (int i = 0; i < num_items; ++i) {
+      if (svect->is_used(i) && pred_func(i, svect)) output[(*num_selected)++] = i;
+    }
+  }
+#endif
+
+  /** @brief Fills selection vector fSelected with element indices passing the user predicate
+   *  @param pred Predicate function taking as arguments the SparseVector pointer and the element index
+   *  @returns Number of selected elements. Must be reset to 0 when calling the function.
+   */
+#ifdef COPCORE_CUDA_COMPILER
+  template <typename Predicate, typename Container>
+  static void select_and_move(Vector_t *svect, Predicate pred_func, Container *output, unsigned *num_selected)
+  {
+    sa_detail::select_and_move_kernel_launcher<Type><<<1, 1>>>((Vector_t *)svect, pred_func, output, num_selected);
```

Comment:
The type of `output` should be `Vector_t` as well, at least that's what `select_and_move_kernel_launcher` expects. Also I don't think you need to cast `svect`

Reply:
Left-overs from the previous implementation...